File: packages/runtime-python/src/eventgen_runtime/adapters/sqs.py
from __future__ import annotations
import json
import os
from eventgen_runtime.types import EventEnvelope, PublishResult


class SqsTransportAdapter:
    def __init__(self, queue_url: str, region: str | None = None) -> None:
        self._queue_url = queue_url
        self._region = region or os.environ.get("AWS_REGION", "us-east-1")

    async def publish(self, envelope: EventEnvelope) -> PublishResult:
        import aioboto3  # type: ignore[import]
        session = aioboto3.Session()
        fifo = envelope.metadata.fifo
        extra: dict = {}
        if fifo:
            extra["MessageGroupId"] = fifo.message_group_id
            if fifo.deduplication_id:
                extra["MessageDeduplicationId"] = fifo.deduplication_id
        async with session.client("sqs", region_name=self._region) as client:
            response = await client.send_message(
                QueueUrl=self._queue_url,
                MessageBody=json.dumps(envelope.__dict__, default=str),
                MessageAttributes={
                    "eventId": {"DataType": "String", "StringValue": envelope.event_id},
                    "version": {"DataType": "String", "StringValue": envelope.version},
                },
                **extra,
            )
        return PublishResult(message_id=response["MessageId"])

File: event-sdk-generator/packages/runtime-go/adapters/sqs.go
package adapters

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/sqs"
	"github.com/aws/aws-sdk-go-v2/service/sqs/types"
	runtime "github.com/eventgen/runtime-go"
)

type SqsTransportAdapter struct {
	client   *sqs.Client
	queueURL string
}

func NewSqsTransportAdapter(ctx context.Context, queueURL string, optFns ...func(*config.LoadOptions) error) (*SqsTransportAdapter, error) {
	cfg, err := config.LoadDefaultConfig(ctx, optFns...)
	if err != nil {
		return nil, fmt.Errorf("sqs adapter: load config: %w", err)
	}
	return &SqsTransportAdapter{client: sqs.NewFromConfig(cfg), queueURL: queueURL}, nil
}

func NewSqsTransportAdapterWithClient(client *sqs.Client, queueURL string) *SqsTransportAdapter {
	return &SqsTransportAdapter{client: client, queueURL: queueURL}
}

func (a *SqsTransportAdapter) Publish(ctx context.Context, envelope runtime.EventEnvelope) (*runtime.PublishResult, error) {
	body, err := json.Marshal(envelope)
	if err != nil {
		return nil, fmt.Errorf("sqs adapter: marshal envelope: %w", err)
	}
	input := &sqs.SendMessageInput{
		QueueUrl:    aws.String(a.queueURL),
		MessageBody: aws.String(string(body)),
		MessageAttributes: map[string]types.MessageAttributeValue{
			"eventId": {DataType: aws.String("String"), StringValue: aws.String(envelope.EventID)},
			"version": {DataType: aws.String("String"), StringValue: aws.String(envelope.Version)},
		},
	}
	if f := envelope.Metadata.Fifo; f != nil {
		input.MessageGroupId = aws.String(f.MessageGroupId)
		if f.DeduplicationId != "" {
			input.MessageDeduplicationId = aws.String(f.DeduplicationId)
		}
	}
	out, err := a.client.SendMessage(ctx, input)
	if err != nil {
		return nil, fmt.Errorf("sqs adapter: send %s: %w", envelope.EventID, err)
	}
	return &runtime.PublishResult{MessageID: aws.ToString(out.MessageId)}, nil
}
